WordPress/wordpress-develop
# Block Bindings: Add context needed by block bindings sources during its processing

Trac ticket: https://core.trac.wordpress.org/ticket/61642

## What?
Block bindings sources sometimes need to access some context that is not initially available in the block. Right now, we are extending it when the source is registered. This pull request moves the logic so it is added when the block bindings are processed. This means:
* Before: Extend context during block bindings source registration. This implies that the blocks extend the context even when they don't use bindings.
* After: Extend context during block bindings processing. This implies that the context is extended ONLY for the blocks where bindings are defined.

## How?
Instead of using the `get_block_type_uses_context` created [back then](https://github.com/WordPress/wordpress-develop/pull/6079) for this use case, use the `available_context` in the block class during bindings processing.

I assume it is not possible to remove that filter because of backward-compatibility.

### Testing instructions

0. Register a custom field in your site to test it. You can use a code snippet like this:
```
register_meta(
	'post',
	'url_custom_field',
	array(
		'show_in_rest' => true,
		'single'       => true,
		'type'         => 'string',
		'default'	   => 'https://wpmovies.dev/wp-content/uploads/2023/04/goncharov-poster-original-1-682x1024.jpeg',
	)
);
```

1. Go to a page and insert an image block.
2. Go to the Code editor and connect the image to the custom field by adding the metadata bindings property. It should look something like this:
```

<figure class="wp-block-image size-large"><img src="https://wpmovies.dev/wp-content/uploads/2023/04/goncharov-poster-original-1-682x1024.jpeg" alt="Alt content" class="wp-image-134"/></figure>

```
3. Save the page and check that in the frontend, the URL defined in the custom field is used.
4. Repeat the process for paragraph, heading, and button blocks.
